fluent.runtime/fluent/runtime/fallback.py

```diff
@@ -0,0 +1,83 @@
+# -*- coding: utf-8 -*-
+
+import codecs
+import os
+
+
+class FluentLocalization(object):
+    def __init__(
+        self, locales, resource_ids, resource_loader,
+        use_isolating=False,
+        bundle_class=None, functions=None,
+    ):
+        self.locales = locales
+        self.resource_ids = resource_ids
+        self.resource_loader = resource_loader
+        self.use_isolating = use_isolating
+        if bundle_class is None:
+            from fluent.runtime import FluentBundle
+            self.bundle_class = FluentBundle
+        else:
+            self.bundle_class = bundle_class
+        self.functions = functions
+        self._bundle_cache = []
+        self._bundle_it = self._iterate_bundles()
+
+    def format_value(self, id, args=None):
+        for bundle in self._bundles():
+            if not bundle.has_message(id):
+                continue
+            msg = bundle.get_message(id)
+            if not msg.value:
+                continue
+            val, errors = bundle.format_pattern(msg.value, args)
+            return val
+        return id
+
+    def _bundles(self):
+        bundle_pointer = 0
+        while True:
+            if bundle_pointer == len(self._bundle_cache):
+                try:
+                    self._bundle_cache.append(next(self._bundle_it))
+                except StopIteration:
+                    return
+            yield self._bundle_cache[bundle_pointer]
+            bundle_pointer += 1
+
+    def _iterate_bundles(self):
```

Comment:
Perhaps I'm missing another layer of this, but it looks to me like the `resource_loader.resources` method should just return an iterator of resources, whereas it's currently returning an iterator containing a single list. So I'd have the `resources()` method of the resource loader just return the list, and then write this like:

```python
    def _iterate_bundles(self):
        for i, locale in enumerate(self.locales):
            bundle = self.bundle_class(self.locales[i:],
                                       functions=self.functions,
                                       use_isolating=self.use_isolating)
            for resource in self.resource_loader.resources(locale, self.resource_ids):
                bundle.add_resource(resource)

            yield bundle
```

Reply:
I made the design more obvious here by making `FluentResourceLoader` be both single and multi-root from the get go.